### ISSUE TITLE
Simply ignore repeated points may lose the chance to get better result.

### ISSUE THREAD

**Issue by donglinjy** [bouthilx/orion.algo.hyperband#2]
I did not test this yet, but per the code, seems for each bracket in the first rung when we sample new points, we will ignore the the points already showed in previous trails, and seems we did not take fidelity dimension into consideration for the similarity comparison. 

1) The first bracket sampled like n points.
2) The second bracket first samples like m points and some of them is repeated comparing with first bracket n points. We removed these repeated ones here, but what if even in the first rung of second bracket, the final used points get bad result metric (loss/accuracy, etc.) compared with the removed ones. So the removed points lose the opportunity to run in the second rung in second bracket which may turns out have better result.

    def sample(self, num, bracket, buffer=10):
        samples = self.space.sample(num * buffer, seed=bracket.seed)

        i = 0
        points = []
        while len(points) < num and i < num * buffer:
            point = samples[i]
            if self.get_id(point) not in self.trial_info:
                point = list(point)
                point[self.fidelity_index] = bracket.rungs[0]['resources']
                points.append(tuple(point))
            i += 1

        if len(points) == 0:
            raise RuntimeError(
                'Hyperband keeps sampling already existing points. This should not happen, '
                'please report this error to '
                'https://github.com/bouthilx/orion.algo.hyperband/issues')

        return points

**Comment by bouthilx** [comment]
Good point. I don't remember reading anything about this in the paper, about how should we handle this. My understanding is that they do not filter like I did. Perhaps we should not filter indeed. This would no affect real distributions because very similar points may get accepted in the different brackets, but for categorical dimensions that may be an issue if some categories lead to _slow learners_.
I would be inclined to stay as close as possible to original paper and remove the filtering.

**Comment by donglinjy** [comment]
I think it is reasonable to not run duplicated trials if all the hyper-parameters are the same, including the fidelity/resource value, but may keep them as a candidate for next rung promotion, if they are better than the new trails, then topK selection will increase the fidelity for next decision.

**Comment by bouthilx** [comment]
Thinking about it, I realize the current implementation is actually OK with regards to this issue. The brackets have their first rung sampled in reversed order, that means brackets with first rung of higher fidelity are first sampled. If a latter bracket sample a duplicate (only considering hyperparameters and ignoring fidelity), it means it already exist in another bracket with a **higher** fidelity, therefore there is no reason we should include it again in another bracket with a lower fidelity.

**Comment by donglinjy** [comment]
But there is no guarantee that the previous higher fidelity is the largest fidelity( or R), like below, in the first rung where s=4, i=0 and s=3, i=0, if there is duplicate samples, if ignore no matter in brackets s=4 or s=3, there is no guarantee that the ignored ones ever reach s=4, i=4 or s=3, i=3.
<img width="448" alt="Screen Shot 2020-02-19 at 1 30 21 PM" src="https://user-images.githubusercontent.com/10850941/74804758-24cfa880-531c-11ea-9991-5969e14fa71d.png">

**Comment by bouthilx** [comment]
Yes, but not if you sample in the reversed order (starting from s=0). If a trial is in bracket 0, we know for sure it runs with budget R, and for bracket 1, we know for sure it runs until r_i=27 at minimum, for bracket 2 until r_i=9, and so on. So, if a configuration is in bracket s=0, we should not sample it again in s > 0, if in bracket s=1, we should not sample it again in s > 1, and so on.

**Comment by donglinjy** [comment]
yeah,  but consider this case, in s=2, i=0,  assume there is one duplicated sample A same as s=1, i=0, A has run r_i=27 but not run r_i=81 as only 2 run r_i=81 in s=1, so we simply remove A in s=2, i=0, what if A with r_i=27 is better than those 3 samples with r_i=27 in s=2, i=1.

**Comment by bouthilx** [comment]
Ok I see your point, maybe by chance the same configuration would pass the threshold for promotion in a bracket and not in another. I would say that this is not the purpose of the brackets, which aim to protect from aggressive allocations. From end of page 8 in the [paper](https://arxiv.org/pdf/1603.06560.pdf):
> By doing so, Hyperband is able to exploit situations in which adaptive allocation works well, while protecting itself in situations where more conservative allocations are required.

In my opinion allowing for duplicates goes against the goal of efficiently allocating ressources. If we already allocated a configuration with a large r_i, it should not be allocated a smaller r_j and rerun twice until r_i just to avoid _unlucky_ population sample. I would rely instead on repetitions of Hyperband for such an issue.

**Comment by donglinjy** [comment]
When we visit the brackets in reversed order, the chance to lose the good ones is reduced. I agree the description of bracket purpose, I was no thinking re-run the trails with smaller r_j again but more like to keep those as candidates for next promotion without re-run if possible.

Also as in another issues mentioned, when we repeat Hyperband execution indefinitely, the reverse only happens in one execution.

**Comment by bouthilx** [comment]
But if we transfert trials down to another bracket to let it be promoted there if the population sample favored it (while the current bracket did not favor it, thus no local promotion), then we need to adjust for the rung sizes, and limit samples accordingly. This would also decrease the diversity of samples across brackets (although maybe only slightly).

Also, what does it mean if it is lucky in a lower bracket but not in the top one? That means at epoch X, it was worst than Y other trials in the top bracket. X being large by definition, since it's in an upper bracket, I doubt having it promoted at epoch X in a lower bracket would yield better results. The Y other trials are likely to be the best anyway.

There is also an issue of traceability. Currently we can trace-back in which bracket the trial falls based on the sequence of fidelity found for a given hyperparameter setting. If we share them across brackets then we break this mechanism. This is a technical detail and there must be other solutions, but I'm not sure it is worth the effort.

**Comment by donglinjy** [comment]
Discussed offline, for single execution of hyperband, we can keep the current way as it already reduce the chance where good candidates not be promoted. But for the repeat execution of hyperband, there will still be some issues which we will look at it when we start working on that.
